### PR TITLE
[labs/ssr] fix server templates dropping top level `td` elements

### PR DESCRIPTION
**Pull request by AndrewJakubowicz**
Fixes: https://github.com/lit/lit/issues/4513

### Context

Using `parse` from parse5 everywhere for server templates is not ideal, and unfortunately has consequences for some cases such as the issue template containing a top level `td` element. Parsing this as a page level HTML causes the element to be removed (which matches browser behavior). However, we instead want to treat this case as a fragment.

### Fix

This is a possible fix. I'm not super stoked about it but it does improve the situation.

Basically, there are 3 tags that I could find that `parseFragment` will not handle as a top-level element: `html`, `head`, and `body`.
Thus, detect those cases and only use `parse` if one of these top-level tags is detected. Otherwise use `parseFragment`.

### Test plan

Added a regression test for the template flagged in the issue.

### Risk

There's technically some other edge cases that can fall out of this. For example, the following template will not work correctly due to the HTML comment throwing off the regex:

```js
html`<!-- My nifty comment --><head>...</head>`
```

This might be acceptable as the situation has been improved by this PR.

### CODE REVIEW

**Comment by justinfagnani** [comment]
I suspect there's a bit more we can do here with more advanced parts of the parse5 API. Looking at the implementation of `parseFragment()` it seems like we can get at the root node and see attributes on it:

```ts
import {Parser} from 'parse5';

const parser = Parser.getFragmentParser();
parser.tokenizer.write('<html foo><div></div></html>', true);

// This has the `foo` attribute
console.log(parser.treeAdapter.getFirstChild(parser.document));

// This only has the <div>
console.log(parser.getFragment());
```

I'm not sure yet how to tie this into a fix, but it seems like more of the information is there internally.

**Comment by AndrewJakubowicz** [comment]
I agree Justin! I'll take a look a little more deeply into what parse5 gives us tomorrow.

**Comment by AndrewJakubowicz** [comment]
@justinfagnani I think an inexpensive regex approach is probably going to be the best to capture the couple page level elements. Using `parseFragment` specifically on templates containing a top level `<html>` doesn't work well (even if manually rigged up) because parse5 removes the `sourceCodeLocation`. I haven't found a way to populate the source code location, which is important because we use it to `flushTo` and `skipTo` around Lit markers.

**Review by augustjk — APPROVED** [review]
Unless we unship the `parse()` change to regress on the page level element support, I like this is a strict improvement of what we currently have.

Trying to not use `parse()` all together could be explored on top of this change too.

File: packages/labs/ssr/src/lib/render-value.ts [hunk 1]
```diff
@@ -223,6 +223,14 @@ type Op =
   | CustomElementClosedOp
   | PossibleNodeMarkerOp;
 
+/**
+ * Any of these top level tags will be removed by parse5's `parseFragment` and
+ * will cause part errors if there are any part bindings. For only these tags,
+ * we use a page-level template `parse`.
+ */
+const REGEXP_TEMPLATE_STARTS_WITH_PAGE_TAG =
+  /^\s*(<(!doctype|html|head|body))/i;
```
Comment (justinfagnani) [inline packages/labs/ssr/src/lib/render-value.ts:232]: I think this will handle leading comments too:
```suggestion
  /^(\s*|<!--.*-->)*(<(!doctype|html|head|body))/is;
```

**Comment by justinfagnani** [comment]
@AndrewJakubowicz thanks for looking into it!

I think the limitations of using non-fragment should be documented along with the tag. I modified the regex to handle comments (I think) so maybe there wouldn't be any to mention.

File: packages/labs/ssr/src/lib/render-value.ts (hunk 1 above)
Comment (AndrewJakubowicz) [reply packages/labs/ssr/src/lib/render-value.ts]: Great suggestion. I've updated regex and added a bunch of tests. Things look pretty great with the comment regex addition.

File: packages/labs/ssr/src/lib/render-value.ts [hunk 2]
```diff
@@ -302,16 +307,22 @@ const getTemplateOpcodes = (
   );
 
   const hydratable = isHydratable(result);
+  const isServerTemplate = !hydratable;
```
Comment (augustjk) [inline packages/labs/ssr/src/lib/render-value.ts:310]: Nit: the rest of the code uses `hydratable` directly as basically meaning server template. e.g. https://github.com/lit/lit/blob/9c02b3876dc927c6a82b4420411256ecbb47c08c/packages/labs/ssr/src/lib/render-value.ts#L460-L470

I don't know why the naming was chosen as such then, but I don't see a need to introduce another boolean, even though I much prefer the name `isServerTemplate`.

**Review by augustjk — APPROVED** [review]
Thank you!

File: .changeset/tough-crews-cover.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'@lit-labs/ssr': patch
+---
+
+Fix a bug where a top level 'td' tag would be removed from a server template.
```
Comment (augustjk) [inline .changeset/tough-crews-cover.md:5]:
```suggestion
Fix a bug where server templates with attribute bindings on certain element tags like `<td>` would throw an error during server render.
```

File: packages/labs/ssr/src/lib/render-value.ts (hunk 2 above)
Comment (AndrewJakubowicz) [reply packages/labs/ssr/src/lib/render-value.ts]: Done!

File: .changeset/tough-crews-cover.md (hunk 3 above)
Comment (AndrewJakubowicz) [reply .changeset/tough-crews-cover.md:5]: Should the fact that it's a top-level tag be mentioned?

```suggestion
Fix a bug where server templates with attribute bindings on certain top-level element tags like `<td>` would throw an error during server render.
```

Comment (augustjk) [reply .changeset/tough-crews-cover.md:5]: Oh yes! I somehow dropped the top-level part as i was rewording. The placement makes it a bit ambiguous though. How do you like
```suggestion
Fix a bug where server templates with attribute bindings on certain element tags like `<td>` used top-level would throw an error during server render.
```

Comment (AndrewJakubowicz) [reply .changeset/tough-crews-cover.md:5]: Love it!

**Review by justinfagnani — APPROVED** [review]
